### PR TITLE
Added Model for Time Series Classification

### PR DESCRIPTION
**Pull request by codeboy5**
I have added the code for a basic RNN Model for the task of time series classification.

```Julia
> data, blocks = load(datarecipes()["ecg5000"]);
> task = TSClassificationSingle(blocks, data);
> model = FastAI.taskmodel(task);
> traindl, validdl = taskdataloaders(data, task, 32);
> callbacks = [ToGPU(), Metrics(accuracy)];
> learner = Learner(model, tasklossfn(task); data=(traindl, validdl), optimizer=ADAM(), callbacks = callbacks);
> fitonecycle!(learner, 10, 0.033)
```

image

As I discussed with @darsnack, the idea is to add an encoding to do the reshaping to (features, batch, time) instead of doing it inside on the RNN Model. Working on that right now.

Have remove the type from StackedLSTM as it was redundant.

### CODE REVIEW

**Comment by codeboy5** [comment]
I have also added a multivariate classification dataset and verified the training loop is working fine for it.

File: FastTimeSeries/src/models/RNN.jl
Comment (ToucheSir) [inline FastTimeSeries/src/models/RNN.jl:29]:
```suggestion
    return RNNModel(recbackbone, Dense(recout, outsize))
```

File: FastTimeSeries/src/models.jl
Comment (ToucheSir) [inline FastTimeSeries/src/models.jl:14]:
```suggestion
    return Models.RNNModel(backbone, outsize = length(outblock.classes), recout = size(output, 1))
```

**Review by ToucheSir — COMMENTED** [review]
Couple of drive-by comments on the WIP here

**Comment by codeboy5** [comment]
Hey guys, @ToucheSir @darsnack @lorenzoh. I have updated the notebook. This PR essentially contains code to load regression datasets and the notebook for the classification task.
I will start adding the Inception Time Model in the next PR as we discussed during our weekly call.

File: FastTimeSeries/src/models/StackedLSTM.jl
Comment (ToucheSir) [inline FastTimeSeries/src/models/StackedLSTM.jl:23]: The `chain =` here can be a `return`.

File: FastTimeSeries/src/models/RNN.jl
Comment (ToucheSir) [inline FastTimeSeries/src/models/RNN.jl:34]: is it intentional to have the initial state trainable? If not, I would add `ChainRulesCore` as a dependency and wrap the `reset!` call like so:
```suggestion
    ChainRulesCore.ignore_derivatives() do
        Flux.reset!(m.recbackbone)
    end
```

File: FastTimeSeries/src/models/RNN.jl
Comment (ToucheSir) [inline FastTimeSeries/src/models/RNN.jl:4]:
```suggestion
tabular2rnn(X::AbstractArray{<:AbstractFloat, 3}) = permutedims(X, (1, 3, 2))
```

File: FastTimeSeries/src/models.jl
Comment (ToucheSir) [inline FastTimeSeries/src/models.jl:27]: Is this one-line code pattern used elsewhere? Otherwise it wouldn't hurt to add some newlines.
```suggestion
@testset "blockbackbone" begin
    @test_nowarn FastAI.blockbackbone(TimeSeriesRow(1,140))
end
```

File: FastTimeSeries/src/models.jl
Comment (ToucheSir) [inline FastTimeSeries/src/models.jl:11]:
```suggestion
```
Not relevant any more?

File: FastTimeSeries/src/models.jl
Comment (ToucheSir) [inline FastTimeSeries/src/models.jl:10]: Should be using [`Flux.outputsize`](https://fluxml.ai/Flux.jl/stable/utilities/#Flux.outputsize) (I think we talked about this in a meeting). Changes required should be minimal :)

File: FastTimeSeries/Project.toml
Comment (ToucheSir) [inline FastTimeSeries/Project.toml:16]:
```suggestion
```
I don't see Zygote used anywhere, so one less dependency doesn't hurt.

File: FastTimeSeries/Project.toml
Comment (ToucheSir) [inline FastTimeSeries/Project.toml:27]:
```suggestion
```
Same as above.

**Review by ToucheSir — COMMENTED** [review]
Looks good to go modulo a couple small things!

**Comment by lorenzoh** [comment]
Great progress in my absence and cool to see the training working!

Are the outstanding comments being addressed in #256 or could this be merged individually beforehand?

File: FastTimeSeries/src/models.jl
Comment (codeboy5) [reply FastTimeSeries/src/models.jl:27]: I think I saw it somewhere, but yeah the new lines look better.

**Comment by codeboy5** [comment]
> Great progress in my absence and cool to see the training working!
> 
> Are the outstanding comments being addressed in #256 or could this be merged individually beforehand?

I am resolving the outstanding comments in #256 , some of them have been resolved and a couple are left, which I will resolve by end of day. This can be merged then I believe.

File: FastTimeSeries/Project.toml
Comment (codeboy5) [reply FastTimeSeries/Project.toml:16]: I tried doing that, but it throws an error 
```julia
ERROR: LoadError: ArgumentError: Package FastTimeSeries does not have Zygote in its dependencies:`
```

File: FastTimeSeries/src/models.jl
Comment (codeboy5) [reply FastTimeSeries/src/models.jl:10]: I tried doing that, but seems to give an error. Will try to debug it later.
https://pastebin.com/mB7aYvdD

File: FastTimeSeries/Project.toml
Comment (ToucheSir) [reply FastTimeSeries/Project.toml:16]: This really could use a full stack trace, because AFAICT you're not using Zygote anywhere? Did you `] resolve` after removing it from the project?

File: FastTimeSeries/src/models.jl
Comment (ToucheSir) [reply FastTimeSeries/src/models.jl:10]: I see, this is https://github.com/FluxML/Flux.jl/pull/1755#issue-1035506488. Can you do the following then?
- Use `zeros` or `ones` instead of `rand`.
- Pass a batch size of 1 instead of 32 and seq length of 1 instead of `inblock.obslength`. This will save some time and you don't need it because you only care about the first output dimension.
- `reset!` the backbone after you calculate `output`.

File: FastTimeSeries/Project.toml
Comment (codeboy5) [reply FastTimeSeries/Project.toml:16]: seems to work now, might have missed the `resolve`. thanks :)

File: FastTimeSeries/src/models.jl
Comment (codeboy5) [reply FastTimeSeries/src/models.jl:10]: Yes, sounds great.

**Comment by ToucheSir** [comment]
I wasn't aware #256 was _also_ working on the StackedRNN model. It would be preferable to merge those changes here so that PR stays more focused on the InceptionTime. If not, we can just close this one and focus on that one (I'm interpreting your comment to mean that the changes in #256 completely subsume this PR).

**Comment by ToucheSir** [comment]
You'd likely have to rebase #256 after merging this PR if you've made changes to the same functions/types in that one. If that doesn't sound like a problem, we can merge this first.

**Review by ToucheSir — APPROVED** [review]
Merging as noted to follow up in #256.

**Pull request merged**
